From 298f15386ff52af3b37185767b76161d1a108e09 Mon Sep 17 00:00:00 2001
From: acierniakx <acierniakx@habana.ai>
Date: Fri, 18 Oct 2024 13:39:37 +0300
Subject: [PATCH] [SW-198841] Check for batch_norm_inf

This check is temporary solution to fix fsdp unit test. Some batch_norm
ops are using BatchNormOpBackend which has not reshapes removed yet. Inference
mode scenario has to be able to handle "reshape" and "no-reshape" variants.

Change-Id: I723a0c16d7337726498cd6740ced6ee22d96afa7
---

diff --git a/hpu_ops/backend/batch_norm_gen.cpp b/hpu_ops/backend/batch_norm_gen.cpp
index 3003168..ea08257 100644
--- a/hpu_ops/backend/batch_norm_gen.cpp
+++ b/hpu_ops/backend/batch_norm_gen.cpp
@@ -370,6 +370,8 @@
          params.get(),
          params_size});
   } else {
+    // remove this case when all BatchNormBackend would be in no-reshape
+    // variants
     std::optional<sh::tensor> inputStorageOpt;
     const auto [input_4d, input_4d_shape] =
         transform_tensor_to_4d<TENSOR_IDX, SHAPE_IDX>(
@@ -464,16 +466,37 @@
   running_var = cast_if_necessary_or_default(
       &op, graph, running_var_opt, running_var, runningVarStorageOpt);
 
-  bn_out.emplace_back(std::move(
-      OpBackend::BuildNode(
-          &op,
-          graph,
-          {get_guid_with_precision("batch_norm_inf_reshape", op.ScalarType()),
-           {input.syn_t, bias, weight, running_mean, running_var},
-           {{out_shapes[INPUT_IDX], op.ScalarType(), c10::optional<int>(0)}},
-           params.get(),
-           params_size})
-          .at(0)));
+  if (is_no_reshape_op(op)) {
+    bn_out.emplace_back(std::move(
+    OpBackend::BuildNode(
+        &op,
+        graph,
+        {get_guid_with_precision("batch_norm_inf_reshape", op.ScalarType()),
+          {input.syn_t, bias, weight, running_mean, running_var},
+          {{out_shapes[INPUT_IDX], op.ScalarType(), c10::optional<int>(0)}},
+          params.get(),
+          params_size})
+        .at(0)));
+  } else { // remove this case when all BatchNormBackend would be in no-reshape
+           // variants
+    const auto input_dim = input.pt_t.sizes().size();
+    std::optional<sh::tensor> inputStorageOpt;
+    const auto [input_4d, input_4d_shape] =
+        transform_tensor_to_4d<TENSOR_IDX, SHAPE_IDX>(
+            op, graph, input, inputStorageOpt);
+    bn_out.emplace_back(std::move(
+        OpBackend::BuildNode(
+            &op,
+            graph,
+            {get_guid_with_precision("batch_norm_inf", op.ScalarType()),
+             {input_4d, bias, weight, running_mean, running_var},
+             {{input_4d_shape,
+               op.ScalarType(),
+               (input_dim != 4) ? c10::nullopt : c10::optional<int>(0)}},
+             params.get(),
+             params_size})
+            .at(0)));
+  }
 
   bn_out.emplace_back(
       std::move(OpBackend::BuildNode(
